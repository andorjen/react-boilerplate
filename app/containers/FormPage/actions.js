import {
  SUBMIT_FORM,
  SUBMIT_FORM_SUCCESS,
  SUBMIT_FORM_ERROR,
  CHANGE_FORM_INPUT,
} from './constants';

function changeFormInput(input) {
  return {
    type: CHANGE_FORM_INPUT,
    payload: input,
  };
}

function submitForm() {
  return {
    type: SUBMIT_FORM,
  };
}

function submitFormSuccess(data) {
  return {
    type: SUBMIT_FORM_SUCCESS,
    payload: data,
  };
}

function submitFormError(error) {
  return {
    type: SUBMIT_FORM_ERROR,
    payload: error,
  };
}

export { changeFormInput, submitForm, submitFormSuccess, submitFormError };
