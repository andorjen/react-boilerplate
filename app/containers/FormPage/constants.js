const BACKEND_URL = 'http://localhost:3000/api/contents';
const SUBMIT_FORM = 'SUBMIT_FORM';
const SUBMIT_FORM_SUCCESS = 'SUBMIT_FORM_SUCCESS';
const SUBMIT_FORM_ERROR = 'SUBMIT_FORM_ERROR';
const CHANGE_FORM_INPUT = 'CHANGE_FORM_INPUT';
const RELOAD_FORM_PAGE = 'RELOAD_FORM_PAGE';

export {
  BACKEND_URL,
  SUBMIT_FORM,
  SUBMIT_FORM_SUCCESS,
  SUBMIT_FORM_ERROR,
  CHANGE_FORM_INPUT,
  RELOAD_FORM_PAGE,
};
