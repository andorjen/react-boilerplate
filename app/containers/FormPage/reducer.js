import {
  SUBMIT_FORM,
  SUBMIT_FORM_SUCCESS,
  SUBMIT_FORM_ERROR,
  CHANGE_FORM_INPUT,
  RELOAD_FORM_PAGE,
} from './constants';

// The initial state of the FormPage
export const initialState = {
  isSubmitting: false,
  hasSubmitted: false,
  formInput: '',
  content: '',
};

/** All reducers for FormPage */
function formReducer(state = initialState, { type, payload }) {
  switch (type) {
    case CHANGE_FORM_INPUT:
      return { ...state, isSubmitting: false, formInput: payload };

    case SUBMIT_FORM:
      return { ...state, isSubmitting: true, hasSubmitted: false };

    case SUBMIT_FORM_SUCCESS:
      return {
        ...state,
        isSubmitting: false,
        hasSubmitted: true,
        formInput: '',
        content: payload.content,
        error: null,
      };

    case SUBMIT_FORM_ERROR:
      return {
        ...state,
        isSubmitting: false,
        hasSubmitted: true,
        error: payload.error,
        content: '',
      };
    case RELOAD_FORM_PAGE:
      return { ...initialState };

    default:
      return state;
  }
}

export { formReducer };
