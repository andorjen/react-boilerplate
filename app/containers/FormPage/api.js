import axios from 'axios';
import { BACKEND_URL } from './constants';

async function postNewContent(data) {
  const cleanupData = data.trim();
  const res = await axios.post(`${BACKEND_URL}/add`, {
    content: cleanupData,
  });
  return res.data;
}

export default postNewContent;
