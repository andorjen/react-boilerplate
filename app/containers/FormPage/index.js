/*
 * Form Page
 *
 * This is the first thing users see of our App, at the '/' route
 *
 */

import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import { compose } from 'redux';

import styled from 'styled-components';
import { useInjectReducer } from 'utils/injectReducer';
import { useInjectSaga } from 'utils/injectSaga';
import {
  makeSelectFormInput,
  makeSelectIsSubmitting,
  makeSelectHasSubmitted,
  makeSelectError,
  makeSelectContent,
} from './selectors';

import { changeFormInput, submitForm } from './actions';

import NavBar from '../../components/NavBar';
import Button from '../../components/Button';
import { formReducer } from './reducer';
import formSaga from './saga';

const FormPageWrapper = styled.div`
  background-color: rgba(2, 219, 125, 0.7);
  min-height: 100vh;
  padding: 4em;
  text-align: center;
  font-family: 'Open Sans', sans-serif;
`;

const FormWrapper = styled.div`
  background-color: white;
  padding: 1em;
  border-radius: 0.4em;
  width: 60%;
  margin-left: 20%;
  margin-top: 5%;
  font-family: 'Open Sans', sans-serif;
`;

const InputBox = styled.textarea`
  width: 60%;
  height: 8em;
`;

const FormLabelWrapper = styled.div`
  text-align: left;
  width: 60%;
  margin-left: 20%;
  padding-top: 1em;
  padding-bottom: 0.5em;
`;

const InlineBlock = styled.div`
  display: inline-block;
  max-height: 100%;
  margin-left: 1em;
`;

const SuccessMessageContainer = styled.div`
  background-color: rgb(201, 242, 202);
  text-align: center;
  font-size: 0.9em;
  padding: 0.2em;
  border-radius: 0.2em;
  width: 60%;
  margin-left: 20%;
`;
const ErrorMessageContainer = styled.div`
  background-color: rgb(250, 181, 165);
  text-align: center;
  font-size: 0.9em;
  padding: 0.2em;
  border-radius: 0.2em;
  width: 60%;
  margin-left: 20%;
`;

const StyledPara = styled.p`
  font-family: 'Open Sans', sans-serif;
  font-size: 0.9em;
`;

function FormPage({
  hasSubmitted,
  isSubmitting,
  formInput,
  postData,
  changeInput,
  error,
  content,
}) {
  useInjectReducer({ key: 'formReducer', reducer: formReducer });
  useInjectSaga({ key: 'formSaga', saga: formSaga });

  return (
    <div>
      <NavBar />
      <FormPageWrapper>
        <FormWrapper>
          <h2>Add Your Own Text</h2>

          {hasSubmitted && content && (
            <SuccessMessageContainer>
              <StyledPara>
                You have successfully submitted the text, go to
                <a href="http://localhost:3000"> HOME PAGE </a>to view.
              </StyledPara>
            </SuccessMessageContainer>
          )}

          {hasSubmitted && error && (
            <ErrorMessageContainer>
              <StyledPara>
                {error.message}. Please make sure to enter valid texts.
              </StyledPara>
            </ErrorMessageContainer>
          )}

          <form onSubmit={postData}>
            <FormLabelWrapper>
              <label htmlFor="input-content"> Text: </label>
            </FormLabelWrapper>
            <InputBox
              id="input-content"
              onChange={changeInput}
              type="text"
              value={formInput}
              placeholder="input length should be between 1 to 1000"
            />

            <div>
              <InlineBlock>
                <Button
                  text="Submit"
                  disabled={isSubmitting}
                  onClick={postData}
                />
              </InlineBlock>
              {isSubmitting && <InlineBlock>Submitting...</InlineBlock>}
            </div>
          </form>
        </FormWrapper>
      </FormPageWrapper>
    </div>
  );
}

FormPage.propTypes = {
  hasSubmitted: PropTypes.bool,
  isSubmitting: PropTypes.bool,
  formInput: PropTypes.string,
  postData: PropTypes.func,
  changeInput: PropTypes.func,
  error: PropTypes.object,
  content: PropTypes.string,
};

const mapStateToProps = createStructuredSelector({
  formInput: makeSelectFormInput(),
  isSubmitting: makeSelectIsSubmitting(),
  hasSubmitted: makeSelectHasSubmitted(),
  error: makeSelectError(),
  content: makeSelectContent(),
});

function mapDispatchToProps(dispatch) {
  return {
    changeInput: evt => {
      dispatch(changeFormInput(evt.target.value));
    },
    postData: evt => {
      if (evt !== undefined && evt.preventDefault) evt.preventDefault();
      dispatch(submitForm());
    },
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
  memo,
)(FormPage);
