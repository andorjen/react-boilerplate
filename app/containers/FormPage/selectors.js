import { createSelector } from 'reselect';
import { initialState } from './reducer';
// const selectRouter = state => state.router;
const selectForm = state => state.formReducer || initialState;

// const makeSelectLocation = () =>
//   createSelector(
//     selectRouter,
//     routerState => routerState.location,
//   );

const makeSelectFormInput = () =>
  createSelector(
    selectForm,
    globalState => globalState.formInput,
  );

const makeSelectIsSubmitting = () =>
  createSelector(
    selectForm,
    globalState => globalState.isSubmitting,
  );

const makeSelectHasSubmitted = () =>
  createSelector(
    selectForm,
    globalState => globalState.hasSubmitted,
  );

const makeSelectError = () =>
  createSelector(
    selectForm,
    globalState => globalState.error,
  );

export {
  selectForm,
  makeSelectFormInput,
  makeSelectIsSubmitting,
  makeSelectHasSubmitted,
  makeSelectError,
};
