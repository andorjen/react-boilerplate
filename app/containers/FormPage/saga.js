import { call, put, select, takeLatest, delay } from 'redux-saga/effects';
import axios from 'axios';
import {
  SUBMIT_FORM,
  SUBMIT_FORM_ERROR,
  SUBMIT_FORM_SUCCESS,
} from './constants';

import { BACKEND_URL } from '../HomePage/constants';
import { makeSelectFormInput } from './selectors';

/** Make Api Call to backend to post data into database */
function* postNewContentToDatabase() {
  // get data from formInput state, clean up by trimming any white spaces
  const data = yield select(makeSelectFormInput());
  const cleanupData = data.trim();

  try {
    // make api call
    const res = yield call(axios.post, `${BACKEND_URL}/add`, {
      content: cleanupData,
    });
    yield delay(1000); // mimick the wait time for api call here, not necessary

    // if success, pass res data to reducer, and set home page to need reload
    yield put({
      type: SUBMIT_FORM_SUCCESS,
      payload: { content: res.data.content },
    });
  } catch (err) {
    // if fail, pass error to reducer
    yield put({
      type: SUBMIT_FORM_ERROR,
      payload: { error: err.response.data.error },
    });
  }
}

/** performs api call on the latest SUBMIT_FORM action */
function* formSaga() {
  yield takeLatest(SUBMIT_FORM, postNewContentToDatabase);
}

export default formSaga;
