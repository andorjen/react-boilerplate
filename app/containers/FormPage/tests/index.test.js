import React from 'react';
import { render, getByText, fireEvent } from 'react-testing-library';
import 'jest-dom/extend-expect';

import { IntlProvider } from 'react-intl';
import { MemoryRouter, browserHistory } from 'react-router-dom';
import { Provider } from 'react-redux';

// import axios from 'axios';

import FormPage from '../index';
import configureStore from '../../../configureStore';

let store;

beforeAll(() => {
  store = configureStore({}, browserHistory);
});

// cleanup after each test
afterEach(() => {
  jest.clearAllMocks();
});

/** tests for rendering of Form page */
describe('<HomePage />', () => {
  it('should render and match the snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <MemoryRouter>
            <FormPage />
          </MemoryRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render the correct info', () => {
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <MemoryRouter>
            <FormPage />
          </MemoryRouter>
        </IntlProvider>
      </Provider>,
    );
    const text = getByText(container, 'Add Your Own Text');
    const submit = getByText(container, 'Submit');
    expect(text).toBeInTheDocument();
    expect(submit).toBeInTheDocument();
  });

  it('should show submitting status after form submission', () => {
    const postData = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <MemoryRouter>
            <FormPage postData={postData} />
          </MemoryRouter>
        </IntlProvider>
      </Provider>,
    );
    fireEvent.submit(document.querySelector('form'));
    const submission = getByText(
      container,
      'Submitting... please do not refresh page',
    );
    expect(submission).toBeInTheDocument();

    const button = document.querySelector('button');
    expect(button).toHaveAttribute('disabled');
  });
});
