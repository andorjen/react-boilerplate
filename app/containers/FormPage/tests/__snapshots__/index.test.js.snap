// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<div>
  <div>
    <nav
      class="NavBar__Nav-ar0vx-0 haeefk"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
        style="background-color: rgba(140, 71, 245, 0.4); font-weight: bold;"
      >
        Home
      </a>
      <a
        href="/add"
      >
        Add A Card
      </a>
    </nav>
    <div
      class="FormPage__FormPageWrapper-depfhu-0 fLzIjr"
    >
      <div
        class="FormPage__FormWrapper-depfhu-1 edPicO"
      >
        <h2
          class="FormPage__FormHeader-depfhu-2 bDoYQN"
        >
          Add Your Own Text
        </h2>
        <form
          class="FormPage__StyledForm-depfhu-3 bFKBor"
        >
          <div
            class="FormPage__FormLabel-depfhu-4 icXOCo"
          >
            <label
              for="input-content"
              style="font-family: Open Sans;"
            >
               
              Text:
               
            </label>
          </div>
          <textarea
            class="FormPage__FormInput-depfhu-5 iBuQbY"
            id="input-content"
            placeholder="input length should be between 1 to 1000"
            type="text"
          />
          <div>
            <div
              class="Wrapper-sc-5dsnpv-0 gPlEEe"
            >
              <button
                class="Button__StyledButton-sc-1fkoyq4-0 dVtzVy"
              >
                Submit
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
